diff --git a/user-identify/CMakeLists.txt b/user-identify/CMakeLists.txt
index 6b1c4c3d..b5048f29 100644
--- a/user-identify/CMakeLists.txt
+++ b/user-identify/CMakeLists.txt
@@ -76,6 +76,14 @@ else ()
   add_definitions(-DLOG_LEVEL=2)
 endif ()
 
+if (DEFINED TEST_MODE)
+  message("test mode: " ${TEST_MODE})
+  add_definitions(-DTEST_MODE=${TEST_MODE})
+else ()
+  message(FATAL_ERROR "TEST_MODE not defined")
+endif ()
+
+
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC -std=c++23")
 
 set(SRHINO_PLUGIN_FRAMEWORK_BUILD_PATH srhino.com/srhino_plugin_framework/v${SRHINO_PLUGIN_FRAMEWORK_VERSION})
@@ -98,9 +106,9 @@ if (BOOTSTRAP_MODE)
   add_dependencies(bootstrap_mode_test proto)
   add_dependencies(bootstrap_mode_cgo_test proto)
 else ()
-  add_subdirectory(test)
-  add_dependencies(test proto)
-  target_link_libraries(test interface source proto -Wl,--whole-archive srhino_plugin_framework_proto -Wl,--no-whole-archive)
+  # add_subdirectory(test)
+  # add_dependencies(test proto)
+  # target_link_libraries(test interface source proto -Wl,--whole-archive srhino_plugin_framework_proto -Wl,--no-whole-archive)
 endif ()
 
 file(GLOB local_source
diff --git a/user-identify/build.sh b/user-identify/build.sh
index a5c1f50c..eb1510d7 100755
--- a/user-identify/build.sh
+++ b/user-identify/build.sh
@@ -33,8 +33,8 @@ else
     mkdir -p $BUILD_DIR
     cd $BUILD_DIR
 
-    cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_C_COMPILER:FILEPATH=/usr/local/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/local/bin/g++ ../
-    # cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_C_COMPILER:FILEPATH=/usr/local/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/local/bin/g++ ../
+    # cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_C_COMPILER:FILEPATH=/usr/local/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/local/bin/g++ ../
+    cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_C_COMPILER:FILEPATH=/usr/local/bin/gcc -DCMAKE_CXX_COMPILER:FILEPATH=/usr/local/bin/g++ ../
 fi
 
 # 执行构建
diff --git a/user-identify/interface.cc b/user-identify/interface.cc
index f35b4432..7ebb4871 100644
--- a/user-identify/interface.cc
+++ b/user-identify/interface.cc
@@ -11,7 +11,7 @@
 
 #include "proto/plugin.pb.h"
 #include "proto/plugin.pb.validate.h"
-#define PROJECT_VERSION "1.0.7"
+#define PROJECT_VERSION "1.0.8"
 
 Plugin* createInstance(Config* config) { return new Plugin(config); }
 
diff --git a/user-identify/source/plugin.cc b/user-identify/source/plugin.cc
index c5c2687d..8ff5b6d0 100644
--- a/user-identify/source/plugin.cc
+++ b/user-identify/source/plugin.cc
@@ -98,7 +98,244 @@ bool isHtmlRequest(const std::string_view accept) {
   return (accept_l.find("text/html") != std::string::npos);
 }
 
+// continue
+#define MODE_CONTINUE 0
+// 请求header直接响应
+#define MODE_DECODE_HEADER_DIRECT_RESPONSE 1
+// 请求header break
+#define MODE_DECODE_HEADER_BREAK 2
+// 请求header 异步调用，调用结束 continue
+#define MODE_DECODE_HEADER_PAUSE_THEN_CONTINUE 3
+// 请求header 异步调用，调用结束直接响应
+#define MODE_DECODE_HEADER_PAUSE_THEN_DIRECT_RESPONSE 4
+
+#define MODE_DECODE_DATA_WAIT_FOR_BODY 5
+#define MODE_DECODE_DATA_TRY_WAIT_FOR_BODY 6
+#define MODE_DECODE_DATA_DIRECT_RESPONSE 7
+#define MODE_DECODE_DATA_PAUSE_THEN_CONTINUE 8
+#define MODE_DECODE_DATA_PAUSE_THEN_DIRECT_RESPONSE 9
+
+#define MODE_ENCODE_HEADER_DIRECT_RESPONSE 10
+#define MODE_ENCODE_HEADER_BREAK 11 
+#define MODE_ENCODE_HEADER_PAUSE_THEN_CONTINUE 12
+#define MODE_ENCODE_HEADER_PAUSE_THEN_DIRECT_RESPONSE 13 
+
+#define MODE_ENCODE_DATA_WAIT_FOR_BODY 14
+#define MODE_ENCODE_DATA_TRY_WAIT_FOR_BODY 15
+#define MODE_ENCODE_DATA_DIRECT_RESPONSE 16
+#define MODE_ENCODE_DATA_PAUSE_THEN_CONTINUE 17
+#define MODE_ENCODE_DATA_PAUSE_THEN_DIRECT_RESPONSE 18
+
+void modifyHeader(HeaderContext& context, std::string header_modify_value) {
+  context.headers().add("Header-To-Add", "plugin add header");
+  context.headers().remove("Header-To-Remove");
+  auto orig_header_value = context.headers().get("Header-To-Modify");
+  std::string modified_header_value(orig_header_value);
+  modified_header_value.append(header_modify_value);
+  context.headers().set("Header-To-Modify", modified_header_value);
+}
+
 HeaderStatus Plugin::onRequestHeader(HeaderContext& context) {
+#if TEST_MODE == MODE_CONTINUE
+  // 测试对 headers 进行增删改
+  modifyHeader(context, "_plugin_modified_onDecodeHeader_Continue");
+  return HeaderStatus::Continue;
+#elif TEST_MODE == MODE_DECODE_HEADER_DIRECT_RESPONSE
+  std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = context.headers().create();
+  header->add("content-type", "text/plain");
+  context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                            std::move("DirectResponse onDecodeHeader"));
+  return HeaderStatus::DirectResponse;
+#elif TEST_MODE == MODE_DECODE_HEADER_BREAK || (TEST_MODE >= MODE_DECODE_DATA_WAIT_FOR_BODY && TEST_MODE <= MODE_DECODE_DATA_PAUSE_THEN_DIRECT_RESPONSE)
+  // 将 headers的修改延迟到 onRequestBody
+  request_header_ = &context;
+  context.setBreakBody(false);
+  return HeaderStatus::Break;
+// 配合 decode data 测试
+#elif TEST_MODE == MODE_DECODE_HEADER_PAUSE_THEN_CONTINUE
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    modifyHeader(context, std::string("_plugin_modified_onDecodeHeader_Pause_then_Continue"));
+    context.continued();
+    return false;
+  }, false);
+  return HeaderStatus::Pause;
+#elif TEST_MODE == MODE_DECODE_HEADER_PAUSE_THEN_DIRECT_RESPONSE
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = context.headers().create();
+    header->add("content-type", "text/plain");
+    context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                              std::move("DirectResponse onDecodeHeader"));
+    return false;
+  }, false);
+  return HeaderStatus::Pause;
+#else
+  return HeaderStatus::Continue;
+#endif
+}
+
+DataStatus Plugin::onRequestBody(BodyContext& context) {
+#if TEST_MODE == MODE_DECODE_HEADER_BREAK
+  if (context.hasMoreBody()) {
+    return DataStatus::WaitForBody;
+  }
+  modifyHeader(*request_header_, "_plugin_modified_onDecodeHeader_Break");
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_DECODE_DATA_WAIT_FOR_BODY
+  if (context.hasMoreBody()) {
+    return DataStatus::WaitForBody;
+  }
+  std::string modified_body = "request body modify onDecodeData wait for body";
+  request_header_->headers().set("content-length", std::to_string(modified_body.size()));
+  context.data().set(modified_body);
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_DECODE_DATA_TRY_WAIT_FOR_BODY
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  std::string modified_body = "request body modify onDecodeData try wait for body";
+  request_header_->headers().set("content-length", std::to_string(modified_body.size()));
+  context.data().set(modified_body);
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_DECODE_DATA_DIRECT_RESPONSE
+  std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = request_header_->headers().create();
+  header->add("content-type", "text/plain");
+  context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                            std::move("DirectResponse onDecodeData"));
+  return DataStatus::DirectResponse;
+#elif TEST_MODE == MODE_DECODE_DATA_PAUSE_THEN_CONTINUE
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    modifyHeader(*request_header_, std::string("_plugin_modified_onDecodeData_Pause_then_Continue"));
+    context.continued();
+    return false;
+  }, false);
+  return DataStatus::Pause;
+#elif TEST_MODE == MODE_DECODE_DATA_PAUSE_THEN_DIRECT_RESPONSE
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = request_header_->headers().create();
+    header->add("content-type", "text/plain");
+    context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                              std::move("DirectResponse onDecodeData"));
+    return false;
+  }, false);
+  return DataStatus::Pause;
+#else
+  return DataStatus::Continue;
+#endif
+}
+
+HeaderStatus Plugin::onResponseHeader(HeaderContext& context) {
+#if TEST_MODE == MODE_CONTINUE
+  // 测试对 headers 进行增删改
+  modifyHeader(context, "_plugin_modified_onEncodeHeader_Continue");
+  return HeaderStatus::Continue;
+#elif TEST_MODE == MODE_ENCODE_HEADER_DIRECT_RESPONSE
+  std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = context.headers().create();
+  header->add("content-type", "text/plain");
+  context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                            std::move("DirectResponse onEncodeHeader"));
+  return HeaderStatus::DirectResponse;
+#elif TEST_MODE == MODE_ENCODE_HEADER_BREAK || (TEST_MODE >= MODE_ENCODE_DATA_WAIT_FOR_BODY && TEST_MODE <= MODE_ENCODE_DATA_PAUSE_THEN_DIRECT_RESPONSE)
+  // 将 headers的修改延迟到 onResponseBody
+  response_header_ = &context;
+  context.setBreakBody(false);
+  return HeaderStatus::Break;
+#elif TEST_MODE == MODE_ENCODE_HEADER_PAUSE_THEN_CONTINUE
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    modifyHeader(context, std::string("_plugin_modified_onEncodeHeader_Pause_then_Continue"));
+    context.continued();
+    return false;
+  }, false);
+  return HeaderStatus::Pause;
+#elif TEST_MODE == MODE_ENCODE_HEADER_PAUSE_THEN_DIRECT_RESPONSE
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = context.headers().create();
+    header->add("content-type", "text/plain");
+    context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                              std::move("DirectResponse onEncodeHeader"));
+    return false;
+  }, false);
+  return HeaderStatus::Pause;
+#else
+  return HeaderStatus::Continue;
+#endif
+}
+
+DataStatus Plugin::onResponseBody(BodyContext& context) {
+#if TEST_MODE == MODE_ENCODE_HEADER_BREAK
+  if (!context.hasMoreBody()) {
+    return DataStatus::WaitForBody;
+  }
+  modifyHeader(*response_header_, "_plugin_modified_onEncodeHeader_Break");
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_ENCODE_DATA_WAIT_FOR_BODY
+  if (context.hasMoreBody()) {
+    return DataStatus::WaitForBody;
+  }
+  std::string modified_body = "response body modify onEncodeData wait for body";
+  response_header_->headers().set("content-length", std::to_string(modified_body.size()));
+  context.data().set(modified_body);
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_ENCODE_DATA_TRY_WAIT_FOR_BODY
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  std::string modified_body = "response body modify onEncodeData try wait for body";
+  response_header_->headers().set("content-length", std::to_string(modified_body.size()));
+  context.data().set(modified_body);
+  return DataStatus::Continue;
+#elif TEST_MODE == MODE_ENCODE_DATA_DIRECT_RESPONSE
+  std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = response_header_->headers().create();
+  header->add("content-type", "text/plain");
+  context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                            std::move("DirectResponse onEncodeData"));
+  return DataStatus::DirectResponse;
+#elif TEST_MODE == MODE_ENCODE_DATA_PAUSE_THEN_CONTINUE
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    modifyHeader(*response_header_, std::string("_plugin_modified_onEncodeData_Pause_then_Continue"));
+    context.continued();
+    return false;
+  }, false);
+  return DataStatus::Pause;
+#elif TEST_MODE == MODE_ENCODE_DATA_PAUSE_THEN_DIRECT_RESPONSE
+  if (context.hasMoreBody()) {
+    return DataStatus::TryWaitForBody;
+  }
+  // 执行异步调用
+  timer_ = context.createTimer(std::chrono::milliseconds(500), [&](Timer& timer) {
+    std::unique_ptr<SrhinoPluginFramework::HeaderMap> header = response_header_->headers().create();
+    header->add("content-type", "text/plain");
+    context.setDirectResponse(Utility::HttpCode::Forbidden, std::move(header),
+                              std::move("DirectResponse onEncodeData"));
+    return false;
+  }, false);
+  return DataStatus::Pause;
+#else
+  // if (context.hasMoreBody()) {
+  //   return DataStatus::TryWaitForBody;
+  // }
+  return DataStatus::Continue;
+#endif
+}
+
+#if 0
+HeaderStatus Plugin::onRequestHeader(HeaderContext& context) {
+  LOG(error, "critical error, bootstrap mode!!");
 #ifdef BOOTSTRAP_MODE
   LOG(error, "critical error, bootstrap mode!!");
   deal_flag_ = false;
@@ -392,12 +629,14 @@ DataStatus Plugin::onRequestBody(BodyContext& context) {
   }
   return DataStatus::Continue;
 }
+#endif
 
 TrailerStatus Plugin::onRequestTrailers(TrailerContext& context) {
   // TODO: 在这里处理请求Trailers的逻辑
   return TrailerStatus::Continue;
 }
 
+#if 0
 HeaderStatus Plugin::onResponseHeader(HeaderContext& context) {
   if (!deal_flag_) {
     return HeaderStatus::Continue;
@@ -606,6 +845,7 @@ DataStatus Plugin::onResponseBody(BodyContext& context) {
   }
   return DataStatus::Continue;
 }
+#endif
 
 TrailerStatus Plugin::onResponseTrailers(TrailerContext& context) {
   // TODO: 在这里处理响应Trailers的逻辑
diff --git a/user-identify/source/plugin.h b/user-identify/source/plugin.h
index f0ddf916..0759a1c1 100644
--- a/user-identify/source/plugin.h
+++ b/user-identify/source/plugin.h
@@ -130,4 +130,7 @@ private:
   bool has_jwt_parsed_{false};
 
   SrhinoPluginFramework::Libs::Encode::DecompressorSharedPtr decompressor_{nullptr};
+  SrhinoPluginFramework::HeaderContext* request_header_{nullptr};
+  SrhinoPluginFramework::HeaderContext* response_header_{nullptr};
+  SrhinoPluginFramework::TimerSharedPtr timer_{nullptr};
 };
